File: haskell/src/Day02.hs
module Day02 where

import Data.List.Extra (splitOn)
import Data.Text (strip)
import Debug.Trace (trace)
import Paths_aoc2023 (getDataFileName)
import Text.Printf (printf)

data Trial = Trial {red :: Int, green :: Int, blue :: Int}

day02 :: IO ()
day02 = do
  inputLines <- lines <$> (getDataFileName "day02-input.txt" >>= readFile)
  printf "part 1: %d\n" (part1 inputLines)
  printf "part 2: %d\n" (part2 inputLines)

part1 :: [String] -> Int
part1 lines = sum [trialScore (n, t) | (n, t) <- [parseGame line | line <- lines]]

part2 :: [String] -> Int
part2 lines = sum [product $ trialVals t | (n, t) <- [parseGame line | line <- lines]]

trialVals :: Trial -> [Int]
trialVals t = map ($ t) [red, green, blue]

singleGroupTrial :: String -> Trial
singleGroupTrial txt =
  let tokens = words txt
      n = read $ head tokens
   in case last tokens of
        "red" -> Trial {red = n, green = 0, blue = 0}
        "green" -> Trial {red = 0, green = n, blue = 0}
        "blue" -> Trial {red = 0, green = 0, blue = n}

parseTrial :: String -> Trial
parseTrial txt =
  let groups = splitOn "," txt
      groupsAsTrials = [singleGroupTrial group | group <- groups]
   in Trial
        { red = sum [red g | g <- groupsAsTrials],
          green = sum [green g | g <- groupsAsTrials],
          blue = sum [blue g | g <- groupsAsTrials]
        }

parseGameId :: String -> Int
parseGameId txt =
  let tokens = splitOn " " txt
   in read (tokens !! 1)

maxTrial :: Trial -> Trial -> Trial
maxTrial a b =
  Trial
    { red = max (red a) (red b),
      green = max (green a) (green b),
      blue = max (blue a) (blue b)
    }

isValid :: Trial -> Bool
isValid t = red t <= 12 && green t <= 13 && blue t <= 14

trialScore :: (Int, Trial) -> Int
trialScore (id, t) = if isValid t then id else 0

parseGame :: String -> (Int, Trial)
parseGame txt =
  let tokens = splitOn ":" txt
      gameId = parseGameId $ head tokens
      trialStr = tokens !! 1
      trial = foldl1 maxTrial [parseTrial t | t <- splitOn ";" trialStr]
   in (gameId, trial)

File: python/day02/day02.py
from collections import defaultdict
from math import prod


def part1():
    with open('../../haskell/data/day02-input.txt') as f:
        games = [parse_game(line) for line in f]
        return sum(game_score(*g) for g in games)


def game_score(game_id, trials):
    def is_valid(trial):
        return all((trial['red'] <= 12,
                    trial['green'] <= 13,
                    trial['blue'] <= 14))
    if all(is_valid(t) for t in trials):
        return game_id
    else:
        return 0


def parse_trial(text: str):
    groups = text.split(',')
    group_record = defaultdict(int)

    def parse_group(group_input: str):
        num, color = group_input.strip().split()
        group_record[color] += int(num)

    for group_input in groups:
        parse_group(group_input)
    return group_record


def parse_game(line: str):
    first, rest = line.split(':')
    game_id = int(first.split()[-1])
    trials = [parse_trial(trial) for trial in rest.split(';')]
    return game_id, trials


def product_of_maxes(trials):
    maxes = defaultdict(int)
    for trial in trials:
        maxes['red'] = max(maxes['red'], trial['red'])
        maxes['green'] = max(maxes['green'], trial['green'])
        maxes['blue'] = max(maxes['blue'], trial['blue'])
    return prod(maxes.values())


def part2():
    with open('../../haskell/data/day02-input.txt') as f:
        games = [parse_game(line) for line in f]
        return sum(product_of_maxes(trials) for game_id, trials in games)


if __name__ == '__main__':
    print(f"part 1: {part1()}")
    print(f"part 2: {part2()}")
